src/object/srv_obj.c

```diff
@@ -265,6 +281,151 @@ obj_bulk_bypass(d_sg_list_t *sgl, crt_bulk_op_t bulk_op)
 	}
 }
 
+static int
+bulk_transfer_sgl(daos_handle_t ioh, crt_rpc_t *rpc, crt_bulk_t remote_bulk,
+		  off_t remote_off, crt_bulk_op_t bulk_op, bool bulk_bind,
+		  d_sg_list_t *sgl, int sgl_idx, struct obj_bulk_args *p_arg)
+{
+	struct bio_desc		*biod;
+	struct crt_bulk_desc	bulk_desc;
+	crt_bulk_perm_t		bulk_perm;
+	crt_bulk_opid_t		bulk_opid;
+	crt_bulk_t		local_bulk;
+	unsigned int		local_off;
+	unsigned int		iov_idx = 0;
+	size_t			remote_size;
+	int			rc;
+
+	if (remote_bulk == NULL) {
+		D_ERROR("Remote bulk is NULL\n");
+		return -DER_INVAL;
+	}
+
+	rc = crt_bulk_get_len(remote_bulk, &remote_size);
+	if (rc) {
+		D_ERROR("Failed to get remote bulk size "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	if (remote_off >= remote_size) {
+		D_ERROR("remote_bulk_off %zu >= remote_bulk_size %zu\n",
+			remote_off, remote_size);
+		return -DER_INVAL;
+	}
+
+	if (daos_io_bypass & IOBP_SRV_BULK) {
+		obj_bulk_bypass(sgl, bulk_op);
+		return 0;
+	}
+
+	biod = daos_handle_is_valid(ioh) ? vos_ioh2desc(ioh) : NULL;
+	bulk_perm = bulk_op == CRT_BULK_PUT ? CRT_BULK_RO : CRT_BULK_RW;
+
+	while (iov_idx < sgl->sg_nr_out) {
+		d_sg_list_t	sgl_sent;
+		size_t		length = 0;
+		unsigned int	start;
+		bool		cached_bulk = false;
+
+		/*
+		 * Skip bulk transfer over IOVs with NULL buffer address,
+		 * these NULL IOVs are 'holes' or deduped records.
+		 */
+		while (iov_idx < sgl->sg_nr_out &&
+		       sgl->sg_iovs[iov_idx].iov_buf == NULL) {
+			remote_off += sgl->sg_iovs[iov_idx].iov_len;
+			iov_idx++;
+		}
+
+		if (iov_idx == sgl->sg_nr_out)
+			break;
+
+		if (remote_off >= remote_size) {
+			D_ERROR("Remote bulk is used up. off:%zu, size:%zu\n",
+				remote_off, remote_size);
+			rc = -DER_OVERFLOW;
+			break;
+		}
+
+		local_bulk = bio_iod_bulk(biod, sgl_idx, iov_idx, &local_off);
```

Comment:
Oh, I see what you mean now. Yes, that can reduce bulk transfer. I'll consider to support it later. Thanks.

Reply:
Discussed with Xuezhao offline. We could optimize this as Xuezhao suggested, but the optimization is quite limited, it can only apply to all 4k aligned IOVs. (Please be aware that SPDK IO is 4k aligned, so even if an IOV is less than 4k, it needs whole 4k DMA buffer, so after mapping IOVs into DMA buffer, there are actually implicit 'holes' in the mapped region if any IOV isn't 4k aligned).